unit_testing_guidelines.md

```diff
@@ -0,0 +1,179 @@
+# Unit and Integration Testing guidelines
+
+This document briefly covers testing philosophy with regard to integration and unit tests,
+especially for the Python language and in the context of developing a KBase core server like
+the Execution Engine.
+
+## Unit versus Integration tests
+
+Unit tests cover one module, class, or function, called a code unit from here on out. For example,
+a unit test file might cover the contents of `my_module.py` or more granularly `my_module.MyClass`.
+Code outside the code unit should be excluded from the tests. The exception is "value classes"
+which are classes which primarily hold data and whose behavior is based on that data. Other
+classes required by the unit under test should be mocked out.
+
+This makes unit tests fast and easy to understand, as only the isolated code unit needs to be
+comprehended in order to comprehend test failures.
+
+In contrast an integration test tests that two or more code units work well together. This can
+range from anything between testing two code units' interactions to api-to-DB tests for a server.
+Integration tests are typically much much slower, much more complex, take much more setup code,
+and are harder to understand. Due to this, it is advisable to mimize the number of integration
+tests to the least possible to ensure the various code units work together correctly, and write
+unit tests to cover as much code as possible. In the author's experience, it is usually not
+difficult to write unit tests with 100% coverage for the code unit.
+
+## Mocking dependencies
+
+As previously described, a unit test should only cover a single unit of code. What this means
+is that complex dependencies (e.g. not simple helper functions that may be called from the
+class, not value classes, etc.) need to be mocked out. We do this via inversion of control, or
+dependency injection. That is, if a code unit needs another code unit as a dependency, the
+dependency should *provided to* the code unit rather than *constructed by* the code unit.
+
+For example, consider a toy function that contacts the workspace service:
+
+```
+def get_object_name_from_id(url, token, ref):
+    ws = Workspace(url, token=token)
+    return ws.get_object_info3({'objects': [{'ref': ref}]})['infos'][1]
+```
+
+Note that the same situation may arise in a class that needs to contact the workspace regularly and
+constructs the client in its `__init__` method.
+
+This makes the function difficult to unit test as if run as is, it will contact the workspace
```

Comment:
comma after `unit test`, hyphenate `as-is`

Reply:
added a parenthetical remark to that effect